### PR TITLE
rgw: sync plugin to UCloud UFile

### PR DESCRIPTION
**Pull request by platoxie**
This new plugin named cloud module sync data from RGW to UFile of UCloud. Cloud module can replicate or remove data in UFile according to RGW. Cloud module support other storage interface expansion, so when adding a cloud storage type, we don't need define a new module in RGW.

### CODE REVIEW

**Comment by joscollin** [comment]
@platoxie 
In Ceph we conventionally put the `Signed-off-by` at the end of a commit message. The title of the Commit should be prefixed with the submodule name followed by a short description of the fix. A detailed description should be added as the body of the Commit message. Please see one of my merged PR as an example: https://github.com/ceph/ceph/pull/15648/commits

See this too: https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#1-sign-your-work

You can do:
```
    $git commit --amend.
    Modify your commit and add the `Signed-off-by` manually for this time.
    Then push using -f option again: $git push origin wip-cloud-sync -f
```

From the next time onwards remember to use `-s` when you commit your changes, so that the `Signed-off-by` will be added automatically to your commit, if you configure your git settings properly.
Please see: https://help.github.com/articles/setting-your-username-in-git/
https://help.github.com/articles/setting-your-email-in-git/

**Comment by cbodley** [comment]
@platoxie i haven't had a chance to look closely, but this looks neat. i'm not sure it warrants the extra Custom abstraction layer though - do you see more plugins using that same interface in the near future?

**Comment by platoxie** [comment]
@joscollin ：I will fix it according to your opinion. Tks!

**Comment by platoxie** [comment]
@cbodley : Custom module implement originally s3 & ufile interface vi setting interface type in “--tier-config”，but someone have been finished s3 module：https://github.com/ceph/ceph/pull/14168. so I delete s3 interface from custom module. I am ready to support COS of Tencent and OSS of alibaba in custom module

**Comment by cbodley** [comment]
@platoxie okay, no objections to the CustomAccess stuff

can you help us come up with a strategy to test this? for example, the elasticsearch module has some tests in `src/test/rgw/rgw_multi/tests_es.py` that `src/test/rgw/test_multi.py` runs when you give it `--num-es-zones` and `--es-endpoint`. ideally, we'd have some kind of automated regression testing to make sure we don't break this with future changes to multisite - maybe using a mock server for the ufile api?

we'll also want to include some documentation that covers how to configure a zone for this module, what features are/aren't supported (versioning?), links to ucloud apis, etc.

File: src/rgw/rgw_sync_module_custom.cc [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+public:
+  RGWGetRemoteObjCB():len(0) {}
+  ~RGWGetRemoteObjCB() { data.clear(); }
+  
+  bufferlist& get_data() { return data; }
+  uint64_t get_len() { return len; } 
+  
+  int handle_data(bufferlist& _bl, off_t _ofs, off_t _len) {
+    bufferptr bp(_bl.c_str(), _len);
+    data.push_back(bp);
```
Comment (yehudasa) [inline src/rgw/rgw_sync_module_custom.cc]: @platoxie this wouldn't scale, you'd end up holding the entire object in memory. What you need to do is feed this data to the sender, and have the sender stream out the data.

**Comment by yehudasa** [comment]
@platoxie thank you very much for this, great seeing new sync modules. My major issue with it is that when syncing it ends up reading the entire object into memory and only then does it send it to the remote backend. This might be ok with smaller objects, but objects can be quite large and this wouldn't work. Once that's fixed it'd be great if we could generalize this to work with other cloud services, seems to me that it's pretty close. Note that I won't be around for the next few weeks, but maybe @cbodley can help here.

**Comment by theanalyst** [comment]
@yehudasa suggested the same for the s3 sync plugin as well, https://github.com/ceph/ceph/pull/14168#issuecomment-290564865 I didn't have the time for looking into it back then yet, I'd also like to help in this @cbodley @platoxie

**Comment by platoxie** [comment]
@theanalyst : Do you mean that each third party cloud store sync as a separate module? Instead of a cloud module can support multiple cloud storages?

**Comment by platoxie** [comment]
@yehudasa @cbodley  I fixed the problem that "reading the entire object into memory".  when the sync object equal or less than 16M, the entire object will be read in once time and sent to ufile. when the sync object bigger than 16M , the object will be read and sent in multipart. The object will be devide into several blocks, the size of block is return by ufile multipart initialization interface.

**Comment by platoxie** [comment]
@cbodley  I added OSS (alibaba cloud) sync class. I updated the help doc & test srcipt.
Now the cloud module support Ufile and OSS cloud storage.

**Comment by ZVampirEM77** [comment]
@joscollin Hi, can you review again?

**Comment by joscollin** [comment]
@cbodley @yehudasa @theanalyst 
I'm approving only the changes that I have requested in this PR. I was not keeping track of this PR, due to lack of time (other works in Ceph). So Please do a `Request Changes` for your own reviews, if needed. Thanks.

**Comment by yehudasa** [comment]
@platoxie not sure how I feel about walking around the streaming issue by setting a multipart upload. I'm not familiar with ucloud, but this won't work cleanly when dealing with other cloud providers (s3 multipart upload is stateful, so we'll need to make sure we deal with the clutter and complete it correctly). In any case, I do see you use the synchronous rgw http api that I think might need to go away sooner than later. I'm currently looking into extending the async api to make it easier to stream data out, and hopefully clean up other stuff as well.

**Comment by yehudasa** [comment]
@platoxie see https://github.com/yehudasa/ceph/commits/wip-rest-write-cr for my version of the other sync plugin (that syncs to aws). Note that it streams data vs reading entire objects. It's incomplete, and I aim at taking the configuration and abstractions from your branch.

File: doc/radosgw/rgw_cloud_sync_plugin.rst [hunk 2]
```diff
@@ -7,39 +7,57 @@ RGW sync plugin: cloud sync module
...
   dest_bucket    #[optional] the dest bucket name in ufile.if not specify,the ufile bucket name is rgw bucket name.
+
+
+oss tier-confg
+  cloud_type     #[forcible] oss cloud name: oss (case insensitive)
+  domain_name    #[forcible] address of the oss cloud data
+  public_key     #[forcible] ufile cloud public key
+  private_key    #[forcible] ufile cloud private key
+  bucket_host    #[optional] same as domain_name
+  prefix_bucket  #[optional] the prefix of ufile bucket name,reduce the ufile bucket names conflict.
```
Comment (ZVampirEM77) [inline doc/radosgw/rgw_cloud_sync_plugin.rst]: It is copy paste~ please fix it.

**Comment by platoxie** [comment]
@yehudasa  You used multipart method when syncing data to aws in your branch. why? You have implemented streaming writes(PR #17365)， I think it can replace the whole object writes or multipart writes. so I will modify my code after your PR #17365 to be merged. Is my thought right?

**Comment by yehudasa** [comment]
@platoxie note that the latest is here: https://github.com/ceph/ceph/pull/18932 (just created this PR for reference). I used multipart upload because there's a limit to the object size that you can upload in a single request, so for objects larger than 5GB multipart upload is required. For smaller objects I still use plain uploads as these don't require keeping an upload state, and should be generally faster.
Note that I'm not completely done with the work there. I still want to add some more flexible config. It'd be great if the S3 and UCloud modules could share underlying infrastructure. Not sure if your original abstraction could work out of the box, but we can think of doing something similar that would isolate the different behaviours.

**Comment by platoxie** [comment]
@yehudasa Tks! now I see that streaming writes is the infrastructure. mulitpart writes is also  required when the object's size is huge.

**Comment by platoxie** [comment]
@yehudasa  I think the new module syncing data to other cloud storage should be a module named "cloud".  Using "cloud" module can sync daya to Aws, Ucloud, Ali OSS, Tencent and so on. But not "aws" module, "ucloud" module... what do you think about it?

**Comment by yehudasa** [comment]
@platoxie I'm not sure it completely makes sense, as different cloud providers will have different config methods. We can do something like `cloud:aws`, and `cloud:ucloud`, etc., and have the generic cloud module instantiate the specific implementation.

**Comment by platoxie** [comment]
@yehudasa we can define any args base on txt in "tier-config"，so it can support configs of some different cloud providers , and each cloud sync interface  must know which args need to read.
I think this approach is more scalable than adding new module when adding a new cloud provider.

**Comment by yehudasa** [comment]
@platoxie basically yes. My intention is that any cloud backend will derive from the base cloud module.

**Comment by liewegas** [comment]
I take it this will be subsumed by @yehudasa's work?

**Comment by platoxie** [comment]
@liewegas @yehudasa   PR#18932 created by yehuda only sync to aws. This PR sync to UFile and OSS of alibaba. I want to rewrite this feature base on #18932 when it would be merged into master.

File: src/test/rgw/test-rgw-bucket-oss-sync.sh [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+  sleep 5s
+  s3cmd -c $out_path/$s3cfg mb s3://$bucket_name
+  for i in $(seq 1 10)
+  do
+    if [ $? -eq 0 ]
+      then break;
+    fi
+    sleep 10s
+    s3cmd -c $out_path/$s3cfg mb s3://$bucket_name
+    echo "s3cmd creaet bucket $bucket_name"
```
Comment (joke-lee) [inline src/test/rgw/test-rgw-bucket-oss-sync.sh]: create

**Comment by yehudasa** [comment]
https://github.com/ceph/ceph/pull/18932

**Pull request closed without merging**